libraries/AP_Scripting/lua_bindings.cpp

```diff
@@ -250,11 +250,111 @@ const luaL_Reg AP_Logger_functions[] = {
     {NULL, NULL}
 };
 
+static int get_analog_in(lua_State *L) {
+
+    AP_HAL::AnalogSource *data = hal.analogin->channel(ANALOG_INPUT_NONE);
+
+    if (data == NULL) {
+        return luaL_error(L, "could not assign analog channel");
+    }
+
+    new_AP_HAL__AnalogSource(L);
+    *check_AP_HAL__AnalogSource(L, -1) = data;
+
+    return 1;
+}
+
+static int set_pinMode(lua_State *L) {
+
+    const int args = lua_gettop(L);
+    if (args > 2) {
+        luaL_argerror(L, args, "too many arguments");
+    } else if (args < 2) {
+        luaL_argerror(L, args, "too few arguments");
+    }
+
+    const lua_Integer pin = luaL_checkinteger(L, 1);
+    luaL_argcheck(L, ((pin >= 0) && (pin <= UINT8_MAX)), 2, "pin number out of range");
+
+    const lua_Integer output = luaL_checkinteger(L, 2);
+    luaL_argcheck(L, ((output >= 0) && (output <= HAL_GPIO_OUTPUT)), 2, "output type out of range");
+
+    hal.gpio->pinMode(static_cast<uint8_t>(pin), static_cast<uint8_t>(output));
+
+    return 0;
+}
+
+static int read_pin(lua_State *L) {
+
+    const int args = lua_gettop(L);
+    if (args != 1) {
+        luaL_argerror(L, args, "expected 1 argument");
+    }
+
+    const lua_Integer pin = luaL_checkinteger(L, 1);
+    luaL_argcheck(L, ((pin >= 0) && (pin <= UINT8_MAX)), 2, "pin number out of range");
+
+    bool b = static_cast<bool>(hal.gpio->read(static_cast<uint8_t>(pin)));
+
+    lua_pushboolean(L, b);
+
+    return 1;
+}
+
+static int write_pin(lua_State *L) {
+
+    const int args = lua_gettop(L);
+    if (args > 2) {
+        luaL_argerror(L, args, "too many arguments");
+    } else if (args < 2) {
+        luaL_argerror(L, args, "too few arguments");
+    }
+
+    const lua_Integer pin = luaL_checkinteger(L, 1);
+    luaL_argcheck(L, ((pin >= 0) && (pin <= UINT8_MAX)), 2, "pin number out of range");
+
+    const lua_Integer output = luaL_checkinteger(L, 2);
+    luaL_argcheck(L, ((output >= 0) && (output <= 1)), 2, "output value out of range");
+
+    hal.gpio->write(static_cast<uint8_t>(pin),static_cast<uint8_t>(output));
+
+    return 0;
+}
+
+static int toggle_pin(lua_State *L) {
+
+    const int args = lua_gettop(L);
+    if (args != 1) {
+        luaL_argerror(L, args, "expected 1 argument");
+    }
+
+    const lua_Integer pin = luaL_checkinteger(L, 1);
+    luaL_argcheck(L, ((pin >= 0) && (pin <= UINT8_MAX)), 2, "pin number out of range");
+
+    hal.gpio->toggle(static_cast<uint8_t>(pin));
+
+    return 0;
+}
+
+const luaL_Reg hal_functions[] = {
+    {"analog_pin", get_analog_in},
+    {"pinMode", set_pinMode},
```

Comment:
The mixture of camel and snake bugs me slightly. Could this just be `mode`?

Reply:
it was just to match `set_pinMode` maybe `pin_mode`

Reply:
i have done that